### PR TITLE
Add ratings block

### PR DESCRIPTION
**Pull request by oandregal**
This PR merges the existing [rating block](https://en.support.wordpress.com/wordpress-editor/blocks/ratings-block/) into Jetpack.

- [x] Fix styles in front-end.
- [x] Add PHPDoc.
- [x] Remove spiciness and priciness blocks from server rendering.
- [x] Remove spiciness and priciness blocks from client rendering.
- [x] Rename folder and block to rating-star.
- [x] Remove SVG and CSS related to spiciness and priciness.
- [x] How to include the [missing files](https://github.com/Automattic/jetpack/pull/14091#issuecomment-558132010) in D35777-code

### CODE REVIEW

**Comment by nosolosw** [comment]
For wanderers: I'm a bit confused as to why front-end styles aren't picked up (front-end should look like in the editor). I'll resume this work tomorrow with a fresh mind.

**Comment by nosolosw** [comment]
This is ready to review.

Note that I've deleted all references to `ratings-spiciness` and `ratings-priciness` except in two places: [CSS class names](https://github.com/Automattic/jetpack/pull/14091/files#diff-bce3106a984c728d0be0793a0b5c98d5R2), and symbols (both [SVG](https://github.com/Automattic/jetpack/pull/14091/files#diff-4b7866259d508e34456c620ffa81b9fdR34) and [UTF-8](https://github.com/Automattic/jetpack/pull/14091/files#diff-4b7866259d508e34456c620ffa81b9fdR14)). I've done this because I hope to be able to load or depend on those files in the WordPress.com code for spiciness and priciness. If we prefer just duplicate the code there I'm happy to remove these as well.

File: extensions/blocks/rating-star/editor.js
Comment (simison) [inline extensions/blocks/rating-star/editor.js]: You should use `registerJetpackBlock` to ensure this block is using Jetpack's block availability framework. See example:

https://github.com/Automattic/jetpack/blob/54d177da89d95b24f77f8886bf8061ad7cb51661/extensions/blocks/business-hours/editor.js#L4-L7

Note also how other blocks are structured as `index.js` and `editor.js`

File: extensions/blocks/rating-star/editor.js
Comment (simison) [inline extensions/blocks/rating-star/editor.js]: Should you save the number of stars from `attributes.rating`?

File: extensions/blocks/rating-star/edit.js
Comment (simison) [inline extensions/blocks/rating-star/edit.js]: You should add `jetpack` textdomain to all translation functions:

`__( 'Settings', 'jetpack' )`

**Comment by simison** [comment]
>Note that I've deleted all references to ratings-spiciness and ratings-priciness except in two cases

Adding code comments explaining why they are there but not in block UI would be helpful. Otherwise someone might come and clean them out in the future without realising...

File: extensions/blocks/rating-star/edit.js
Comment (nosolosw) [reply extensions/blocks/rating-star/edit.js]: done at https://github.com/Automattic/jetpack/pull/14091/commits/afc501de25bf66c5b29db80859026cd6e3cf56eb

File: extensions/blocks/rating-star/editor.js
Comment (nosolosw) [reply extensions/blocks/rating-star/editor.js]: done at https://github.com/Automattic/jetpack/pull/14091/commits/afc501de25bf66c5b29db80859026cd6e3cf56eb

File: extensions/blocks/rating-star/editor.js
Comment (nosolosw) [reply extensions/blocks/rating-star/editor.js]: No, the `save` function takes it from the attributes.

File: extensions/blocks/rating-star/icon.js
Comment (simison) [inline extensions/blocks/rating-star/icon.js]: Is there a risk that `props` doesn't exist? You could default it to empty object:

```js
( props: {} ) => { ... }
```

You could also [set a default value when destructuring from an object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Default_values_2), like so:

```js
const { className: '' } = props;
```

File: extensions/blocks/rating-star/editor.js
Comment (simison) [inline extensions/blocks/rating-star/editor.js]: Suppose these labels need `__()`?

File: extensions/blocks/rating-star/editor.js
Comment (simison) [inline extensions/blocks/rating-star/editor.js]: It would be great to have 3 `keywords` here. :-)

See example;
https://github.com/Automattic/jetpack/blob/d001db24166a884881e7d0a2f1c7e12429e053c6/extensions/blocks/tiled-gallery/index.js#L203-L207

File: extensions/blocks/rating-star/style.scss
Comment (simison) [inline extensions/blocks/rating-star/style.scss]: You can pull this from a variable:
```scss
@import '../../shared/styles/jetpack-variables.scss';

margin-bottom: $jetpack-block-margin-bottom;
```

File: extensions/blocks/rating-star/style.scss
Comment (simison) [inline extensions/blocks/rating-star/style.scss]: Not a biggie and ok as-is, but would you prefer to use Gutenberg core variable here to ensure it stays in sync?

If so, you can copy that variable [from Gutenberg](https://github.com/WordPress/gutenberg/blob/2f68f2885135a4da495b1f94978b26e18a394a31/packages/base-styles/_variables.scss#L67) to [`extensions/shared/styles/gutenberg-variables.scss`](https://github.com/Automattic/jetpack/blob/7e8ff7012b98ae5ff4ced1f5f8915974e4652ab1/extensions/shared/styles/gutenberg-variables.scss) and use it from there.

[I'll replace that file soon](https://github.com/Automattic/jetpack/pull/14092) with core's `base-styles` package so things will just stay in sync without manual effort. 🎉

File: extensions/blocks/rating-star/style.scss
Comment (simison) [inline extensions/blocks/rating-star/style.scss]: Would partial match CSS be good here?
```css
[class^=wp-block-jetpack-rating-] { }
```

That would save a lot of CSS being generated. :-)

Comment (nosolosw) [reply extensions/blocks/rating-star/style.scss]: Ended up removing these as per https://github.com/Automattic/jetpack/pull/14091/commits/cc9a951e2b98321aa6307663bab923fd73abba96

File: extensions/blocks/rating-star/editor.js
Comment (nosolosw) [reply extensions/blocks/rating-star/editor.js]: Why three? I can only think of `star` and `rating` as relevant.

Comment (nosolosw) [reply extensions/blocks/rating-star/editor.js]: There is also `review`!

Comment (nosolosw) [reply extensions/blocks/rating-star/editor.js]: added at https://github.com/Automattic/jetpack/pull/14091/commits/8c74d9db5b3f424dbe259533dc170564cda3b883

File: extensions/blocks/rating-star/editor.js
Comment (nosolosw) [reply extensions/blocks/rating-star/editor.js]: yeah, good catch. Done at https://github.com/Automattic/jetpack/pull/14091/commits/8c74d9db5b3f424dbe259533dc170564cda3b883

File: extensions/blocks/rating-star/editor.js
Comment (simison) [reply extensions/blocks/rating-star/editor.js]: 3 is maximum so why not use them all. :-)

Comment (simison) [reply extensions/blocks/rating-star/editor.js]: Block inserter search picks terms from the block's name so you can still replace "star" and "rating" keywords with something else.

Some more ideas; "like", "reaction", "review", "rate", "rank", "vote", "grade"

File: extensions/blocks/rating-star/index.js
Comment (jeherve) [inline extensions/blocks/rating-star/index.js]: You'll need to name it `rating-star` or it won't be made available.

```suggestion
export const name = 'rating-star';
```

File: extensions/blocks/rating-star/rating-meta.php
Comment (jeherve) [inline extensions/blocks/rating-star/rating-meta.php]: I wonder if we should add more to that markup to make the block output more accessible to screen readers? I'm no pro there, I am not quite sure how this would get read. VoiceOver does not seem to read anything in my tests.

**Review by jeherve — CHANGES_REQUESTED** [review]
I'm not sure if that's a core issue, but I see the following when trying to switch styles:

image

File: extensions/blocks/rating-star/icon.js
Comment (nosolosw) [reply extensions/blocks/rating-star/icon.js]: I think this is fine as it is.

File: extensions/blocks/rating-star/style.scss
Comment (nosolosw) [reply extensions/blocks/rating-star/style.scss]: Now we know the final place where ratings will live, I guess is proper to start using the variables defined in that place. Done at https://github.com/Automattic/jetpack/pull/14091/commits/45aac1b6bd9934fc5574db4855a8e7896da158d4

File: extensions/blocks/rating-star/style.scss
Comment (nosolosw) [reply extensions/blocks/rating-star/style.scss]: Done at https://github.com/Automattic/jetpack/pull/14091/commits/a54a0d2d8c44d873deafa55b4a1d97fb5e6e1a1f

File: extensions/blocks/rating-star/rating-meta.php
Comment (nosolosw) [reply extensions/blocks/rating-star/rating-meta.php]: I'd let @jasmussen chime in here, but I think it is ok as it is as per these changes (see issue description) https://github.com/Automattic/wpcom-blocks/pull/123

File: extensions/blocks/rating-star/index.js
Comment (nosolosw) [reply extensions/blocks/rating-star/index.js]: I'm not a native either, but it sounds fine to me as the number can be 1 to anything. The fact that the symbol is a star is an aesthetic choice, what matters is that you're able to rate how do you feel about something.

File: extensions/blocks/rating-star/rating-meta.php
Comment (nosolosw) [reply extensions/blocks/rating-star/rating-meta.php]: In `rating-meta` I purposely didn't use anything Jetpack as to be able to copy it verbatim to where [we bundle](https://github.com/Automattic/wpcom-blocks/pull/128) priciness and spiciness -- which are going to be deprecated.

I'm fine with this file changing overtime, but just for this release I want them both equals, as to be able to potentially retrofit any bugfix that migh pop-up, etc.

File: extensions/blocks/rating-star/index.js
Comment (nosolosw) [reply extensions/blocks/rating-star/index.js]: Fixed at https://github.com/Automattic/jetpack/pull/14091/commits/35310a701e1e98c72febdccfa7f9b6130df90f8f

**Comment by nosolosw** [comment]
> I'm not sure if that's a core issue, but I see the following when trying to switch styles:

@jeherve Could you elaborate what's the issue? I've tested and it seems to work fine.

File: extensions/blocks/rating-star/rating-meta.php
Comment (jasmussen) [reply extensions/blocks/rating-star/rating-meta.php]: Things should actually be a fair bit improved here compared to before. I don't know that the past version was very screen-reader accessible at all (😔) but the new one at least reads aloud the stars as "White Medium Star, White Medium Star, White Medium Star" (for a 2.5 or 3 star rating).

Here's a GIF I recorded trying to learn how the emoji are announced:

emoji

In this one, the 💲 emoji is announced as "Dollar Sign" in comparison to $ which is announced as "dollar". 

I would welcome additional accessibility improvements; I wonder if we could output literal text instead of literal emoji that perhaps read out something like "2.5 White Medium Star rating" or something. 

I would suggest that to be a nice future update to make, but given the improvement over what's there today, I think we should ship this as an iterative improvement.

**Comment by nosolosw** [comment]
In reviewing D35777-code, I've noticed that a few files are missing:

- The `extensions/rating-star` folder, that contains the PHP files that register the block.
- The `_inc/rating-star` folder that contains the view files.

Should I add them manually after this lands (so they don't get overriden) or is there any other thing I should do to register the block?

File: extensions/blocks/rating-star/rating-meta.php
Comment (jeherve) [reply extensions/blocks/rating-star/rating-meta.php]: Ok, that makes sense. Do you think you could create a master issue with the things we'll need to update, or one issue for each thing we want to update? 

This way we'll be able to keep track of the things we want to change.

**Comment by jeherve** [comment]
> Could you elaborate what's the issue? I've tested and it seems to work fine.

The stars are aligned to the left of the preview window, with no padding or anything. I would expect them to be aligned to the center.

**Comment by jeherve** [comment]
> I've noticed that a few files are missing:
> 
> - The extensions/rating-star folder, that contains the PHP files that register the block.
> - The _inc/rating-star folder that contains the view files.

You will indeed need to add the `extensions/rating-star` and all its components manually (php, js, scss, everything from that directory must be committed).
Once you svn add those files, TeamCity will be able to build a new bundle for you.  You consequently don't need to commit the `_inc/blocks/rating-star` directory since TeamCity takes care of creating and adding the built files to your diff for you, and you can then commit everything from the diff the bot will have updated. 

Let's do this together once you're ready if you'd like!

File: extensions/blocks/rating-star/rating-meta.php
Comment (jeherve) [reply extensions/blocks/rating-star/rating-meta.php]: That's interesting, on my end VoiceOver didn't pick up anything when inside a post, but I must have been doing something wrong. I get simple "yellow five-pointed star" indicators when loading our markup on its own:
https://jeherve.keybase.pub/testing-rating.html

I think we could add some content to wrap the whole rating to mention what it is, but I agree that it's not really a blocker for now.

Comment (jeherve) [reply extensions/blocks/rating-star/rating-meta.php]: Logged this here:
https://github.com/Automattic/jetpack/issues/14120

File: extensions/blocks/rating-star/rating-meta.php
Comment (nosolosw) [reply extensions/blocks/rating-star/rating-meta.php]: This is a typo I introduced this morning. Fixed at https://github.com/Automattic/jetpack/pull/14091/commits/93b123b2259aea969a40616a08020d1e6c60b594

File: extensions/blocks/rating-star/rating-meta.php
Comment (nosolosw) [reply extensions/blocks/rating-star/rating-meta.php]: This is already done at [the function that calls this one](https://github.com/Automattic/jetpack/pull/14091/files/77d15f68b8bd0a144420f8c0620fc3ffd8e7f892#diff-4b7866259d508e34456c620ffa81b9fdR53))  as a way to centralize it for them all (star, price, spiciness) - at least for the `color` attribute anyway, which is the only potentially provided by the user.

File: extensions/blocks/rating-star/rating-meta.php
Comment (nosolosw) [reply extensions/blocks/rating-star/rating-meta.php]: Happy to, but not sure which are the things we want to change? This code works fine and takes into account all cases, so using `Jetpack_Gutenberg::block_classes` would be unnecessary.

**Comment by nosolosw** [comment]
> The stars are aligned to the left of the preview window, with no padding or anything. I would expect them to be aligned to the center.

Ah, ok. So, alignment is a user choice and by default they're left-aligned. I agree a bit more of padding could make them stand out a bit more (and also make the symbols bigger) but not sure how we would write CSS that targets the preview without compromising the other scenarios. I'd say this could be a future improvement if we wanted to.

**Review by jeherve — DISMISSED** [review]
This is looking good. I think I'll merge this now, it's in a good state and we can start testing it tomorrow with the Beta!

**Comment by jeherve** [comment]
Just pushed 14ee7e3a7868118350f84c0a9210a58b20fdcbb2 to fix some comments. Will merge when the tests complete.

**Review by jeherve — APPROVED** [review]

**Comment by simison** [comment]
Please remember to commit D35777-code so that files are back to sync.

Nice work on this, @nosolosw ! 🌟

File: extensions/blocks/rating-star/rating-meta.php
Comment (jeherve) [reply extensions/blocks/rating-star/rating-meta.php]: I created #14127 to track this.

While the current implementation works fine, I'd like for things to remain consistent between all our blocks. It also makes it easier to update things later if we ever have to because of a Core update, or decide to change the way we handle classes for all our blocks.

**Comment by nosolosw** [comment]
D35777-code has also been deployed and this is already on WordPress.com

File: extensions/blocks/rating-star/index.js
Comment (akirk) [inline extensions/blocks/rating-star/index.js]: The title of the block cannot be translated. Is this on purpose? Since it has already been referenced in support documents it is used untranslated everywhere.

Comment (jeherve) [reply extensions/blocks/rating-star/index.js]: This was an oversight, nice catch. 

https://github.com/Automattic/jetpack/pull/14190

**Pull request merged**
